### PR TITLE
Issue-1371: Implement Iterable Api

### PR DESCRIPTION
**Pull request by bastikempken**
#### Check List:
* Fixes Issue 1371
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

File: src/main/java/org/assertj/core/internal/Maps.java [hunk 1]
```diff
@@ -556,7 +557,40 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
...
+
+    assertNotNull(info, actual);
+    failIfNull(keys);
+    if (actual.isEmpty() && !keys.iterator().hasNext()) {
+      return;
+    }
+    if (!keys.iterator().hasNext()) {
+      throw new IllegalArgumentException(ErrorMessages.iterableOfKeysToLookForIsEmpty());
+    }
+    final K[] keysAsArray = (K[]) StreamSupport.stream(keys.spliterator(), false).toArray(Object[]::new);
```
[the remark this answers is not in the document]
Comment (bastikempken) [reply src/main/java/org/assertj/core/internal/Maps.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  private static HashSet<String> newHashSet(String entry) {
+    HashSet<String> notExpected = new HashSet<>();
+    notExpected.add(entry);
+    return notExpected;
+  }
+
+  private String[] convertToIterable(Iterable<String> iterable) {
+    return StreamSupport.stream(iterable.spliterator(), false).toArray(String[]::new);
```
[the remark this answers is not in the document]
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    shouldHaveThrown(AssertionError.class);
+  }
+
+  private static HashSet<String> newHashSet(String entry) {
+    HashSet<String> notExpected = new HashSet<>();
+    notExpected.add(entry);
+    return notExpected;
+  }
+
+  private String[] convertToIterable(Iterable<String> iterable) {
```
[the remark this answers is not in the document]
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: of course :)

**Comment by bastikempken** [comment]
@PascalSchumacher thanks for your feedback, i got some valuable input for a better solution on https://github.com/joel-costigliola/assertj-core/issues/1371 . Thats the reason why the implementation became a little bit different now

**Comment by bastikempken** [comment]
@PascalSchumacher Could i do something to improve the code?

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java [hunk 4]
```diff
@@ -962,6 +962,37 @@ public SELF containsOnlyKeys(@SuppressWarnings("unchecked") K... keys) {
...
+   * ringBearers.put(vilya, elrond);
+   * ringBearers.put(oneRing, frodo);
+   *
+   * // assertion will pass
+   * List&lt;Ring&gt; keysToCheck = Arrays.asList(oneRing, nenya, narya, vilya);
+   * assertThat(ringBearers).containsOnlyKeys(keysToCheck);
+   *
+   * // assertion will fail
+   * keysToCheck = Arrays.asList(oneRing, nenya);
+   * assertThat(ringBearers).containsOnlyKeys(keysToCheck);</code></pre>
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractMapAssert.java]: add an example where `keysToCheck` has keys not not in `ringBearers`

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java [hunk 5]
```diff
@@ -962,6 +962,37 @@ public SELF containsOnlyKeys(@SuppressWarnings("unchecked") K... keys) {
...
+   *
+   * @param keys the given keys that should be in the actual map.
+   * @return {@code this} assertions object
+   * @throws AssertionError if the actual map is {@code null}.
+   * @throws AssertionError if the actual map does not contain the given keys, i.e. the actual map contains some or none
+   *           of the given keys, or the actual map contains more entries than the given ones.
+   * @throws IllegalArgumentException if the given argument is an empty array.
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractMapAssert.java]: `is null or empty`

File: src/main/java/org/assertj/core/internal/Maps.java [hunk 6]
```diff
@@ -550,12 +551,37 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
...
+                                            Iterable<K> keys) {
+    final K[] keysAsArray = keys != null ? (K[]) Streams.stream(keys).toArray(Object[]::new) : null;
+    assertContainsOnlyKeys(info, actual, "keys iterable", keysAsArray);
+  }
+
+  private <K, V> void assertContainsOnlyKeys(AssertionInfo info, Map<K, V> actual,
+                                             String placeholderForErrorMessages, K... keys) {
     assertNotNull(info, actual);
-    failIfNull(keys);
+    failIfNull(keys, ErrorMessages.keysToLookForIsNull(placeholderForErrorMessages));
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Maps.java]: static import `keysToLookForIsNull`

File: src/main/java/org/assertj/core/internal/Maps.java [hunk 7]
```diff
@@ -550,12 +551,37 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
...
     if (actual.isEmpty() && keys.length == 0) {
       return;
     }
-    failIfEmpty(keys);
+    failIfEmpty(keys, ErrorMessages.keysToLookForIsEmpty(placeholderForErrorMessages));
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Maps.java]: static import `keysToLookForIsNull` for readability

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+ * @author Sebastian Kempken
+ */
+public class Maps_assertContainsOnlyKeys_iterable_Test extends MapsBaseTest {
+
+  private final String PLACEHOLDER_FOR_ERROR_MESSAGE = "keys iterable";
+
+  @Test
+  public void should_fail_if_actual_is_null() {
+    assertThatExceptionOfType(AssertionError.class).isThrownBy(() -> maps.assertContainsOnlyKeys(someInfo(), null,
+                                                                                                 Arrays.asList("name")))
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: replace all `Arrays.asList` by `org.assertj.core.util.Lists.list` and static import it for readability.

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                                                   .withMessage(actualIsNull());
+  }
+
+  @Test
+  public void should_fail_if_given_keys_array_is_null() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: I think all `_array` can be replaced by `_iterable`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * Tests for
+ * <code>{@link org.assertj.core.internal.Maps#assertContainsOnlyKeys(AssertionInfo, Map, Iterable<Object>)}</code>
+ * .
+ *
+ * @author Sebastian Kempken
+ */
+public class Maps_assertContainsOnlyKeys_iterable_Test extends MapsBaseTest {
+
+  private final String PLACEHOLDER_FOR_ERROR_MESSAGE = "keys iterable";
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: rename to `KEYS_ITERABLE`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+  public void should_fail_if_given_keys_array_is_null() {
+    assertThatNullPointerException().isThrownBy(() -> maps.assertContainsOnlyKeys(someInfo(), actual,
+                                                                                  (Iterable<String>) null))
+                                    .withMessage(keysToLookForIsNull(PLACEHOLDER_FOR_ERROR_MESSAGE));
+  }
+
+  @Test
+  public void should_fail_if_given_keys_array_is_empty() {
+    assertThatIllegalArgumentException().isThrownBy(() -> maps.assertContainsOnlyKeys(someInfo(), actual,
+                                                                                      Arrays.asList(emptyKeys())))
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: `Arrays.asList(emptyKeys())` by `list()` or `Collections.emptyList()` (and static import it)

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                                        .withMessage(keysToLookForIsEmpty(PLACEHOLDER_FOR_ERROR_MESSAGE));
+  }
+
+  @Test
+  public void should_pass_if_actual_and_entries_are_empty() {
+    maps.assertContainsOnlyKeys(someInfo(), emptyMap(), Arrays.asList(emptyKeys()));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: see my previous comment

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+  @Test
+  public void should_fail_if_given_keys_array_is_empty() {
+    assertThatIllegalArgumentException().isThrownBy(() -> maps.assertContainsOnlyKeys(someInfo(), actual,
+                                                                                      Arrays.asList(emptyKeys())))
+                                        .withMessage(keysToLookForIsEmpty(PLACEHOLDER_FOR_ERROR_MESSAGE));
+  }
+
+  @Test
+  public void should_pass_if_actual_and_entries_are_empty() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: rename `entries` to `given_keys`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    maps.assertContainsOnlyKeys(someInfo(), emptyMap(), Arrays.asList(emptyKeys()));
+  }
+
+  @Test
+  public void should_pass_if_actual_contains_only_expected_entries() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: rename `entries` to `keys`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    maps.assertContainsOnlyKeys(someInfo(), actual, Arrays.asList("color", "name"));
+  }
+
+  @Test
+  public void should_fail_if_actual_contains_unexpected_entry() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: `should_fail_if_actual_contains_an_unexpected_key`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+      return;
+    }
+    shouldHaveThrown(AssertionError.class);
+  }
+
+  @Test
+  public void should_fail_if_actual_does_not_contains_every_expected_entries() {
+    AssertionInfo info = someInfo();
+    Iterable<String> expectedKeys = Arrays.asList("name", "color");
+    Map<String, String> underTest = Maps.mapOf(entry("name", "Yoda"));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: static import `mapOf`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    } catch (AssertionError e) {
+      verify(failures).failure(info, shouldContainOnlyKeys(actual, convertToArray(expectedKeys), emptySet(),
+                                                           newHashSet("color")));
+      return;
+    }
+    shouldHaveThrown(AssertionError.class);
+  }
+
+  @Test
+  public void should_fail_if_actual_does_not_contains_every_expected_entries() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: `every_expected_entries` -> `all_expected_keys`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+    } catch (AssertionError e) {
+      verify(failures).failure(info, shouldContainOnlyKeys(underTest, convertToArray(expectedKeys),
+                                                           newHashSet("color"), emptySet()));
+      return;
+    }
+    shouldHaveThrown(AssertionError.class);
+  }
+
+  @Test
+  public void should_fail_if_actual_does_not_contains_every_expected_entries_and_contains_unexpected_one() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: `every_expected_entries` -> `all_expected_keys`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    AssertionInfo info = someInfo();
+    Iterable<String> expectedKeys = Arrays.asList("name", "color");
+    Map<String, String> underTest = Maps.mapOf(entry("name", "Yoda"), entry("job", "Jedi"));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: static import `mapOf`

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests for
+ * <code>{@link org.assertj.core.internal.Maps#assertContainsOnlyKeys(AssertionInfo, Map, Iterable<Object>)}</code>
+ * .
+ *
+ * @author Sebastian Kempken
+ */
+public class Maps_assertContainsOnlyKeys_iterable_Test extends MapsBaseTest {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: some of the tests were not named accurately, the cause being in `Maps_assertContainsOnlyKeys_Test`which you copied (and were right to do so), can you update `Maps_assertContainsOnlyKeys_Test` with the naming improvements I suggested on this file ? (with some variation when needed).

**Comment by joel-costigliola** [comment]
@bastikempken I have reviewed your PR, a bunch of comments but minor ones, thanks for the good work!

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java (hunk 4 above)
Comment (bastikempken) [reply src/main/java/org/assertj/core/api/AbstractMapAssert.java]: done

File: src/main/java/org/assertj/core/api/AbstractMapAssert.java (hunk 5 above)
Comment (bastikempken) [reply src/main/java/org/assertj/core/api/AbstractMapAssert.java]: done

File: src/main/java/org/assertj/core/internal/Maps.java (hunk 6 above)
Comment (bastikempken) [reply src/main/java/org/assertj/core/internal/Maps.java]: done

File: src/main/java/org/assertj/core/internal/Maps.java (hunk 7 above)
Comment (bastikempken) [reply src/main/java/org/assertj/core/internal/Maps.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 8 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 11 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 12 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 13 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 15 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 16 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 17 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 18 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 19 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 20 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

File: src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java (hunk 14 above)
Comment (bastikempken) [reply src/test/java/org/assertj/core/internal/maps/Maps_assertContainsOnlyKeys_iterable_Test.java]: done

**Comment by bastikempken** [comment]
@joel-costigliola i pushed a commit with the changes (squashed with the previous commits), what do you think ? :)

**Comment by joel-costigliola** [comment]
Integrated thanks @bastikempken!
I have changed the testing strategy by simply checking that `containsOnlyKeys(Iterable)` was calling `assertContainsOnlyKeys` with the right array. This will avoid test duplication.

**Pull request closed without merging**